### ISSUE TITLE
A few questions about clustering.

### ISSUE THREAD

**Issue by testark** [vernemq/vernemq#214]
(Not exactly an issue, so please excuse me for opening this.)

I'm in the process of automatic a VerneMQ deployment, and I have a few questions I'd like to ask:
1. When a node, say `vmq@vernemq-1.lan`, leaves the cluster via `vmq-admin cluster leave node="vmq@vernemq-1.lan" -k -i 10 -t 86400` successfully, does this mean that all data has been migrated to other nodes? Is it safe to / should I wipe the `data` directory at this point?
2. Let's say this node has left the cluster successfully and now I want it to rejoin. What kind of precautions should I take concerning the `data` directory? Is it safe to rejoin without changing anything in `data`?
3. Let's say my discovery node is `vmq@vernemq-0.lan`. Why does VerneMQ keep outputting

``` json
{
    "component": "plumtree_broadcast:exchange/423",
    "date": "2016-10-25 13:21:07.816",
    "message": "started plumtree_metadata_manager exchange with 'vmq@vernemq-1.lan' (<0.556.0>)",
    "pid": "<0.141.0>",
    "severity": "debug"
}
```

on this same node after `vmq@vernemq-1.lan` is no longer a part of the cluster?

**Comment by dergraf** [comment]
This is actually a good question, and we should clearify that in our documentation.

The leave command is implemented in a way that it should migrate all queues including its messages. However the `vmq-admin cluster leave` command is time trapped via `-t 86400` and terminates after this time is exceeded. Once this time is exceeded the node has left cluster membership and has shut down. You should see the result of the cluster leave in the `console.log` like:

    MIGRATION SUMMARY: 100 queues migrated, 10000 messages

The code is the best documentation:
```erlang
LeaveFun = fun() ->
    %% stop all MQTT sessions on Node
    %% Note: ensure loadbalancing will put them on other nodes
    vmq_ranch_config:stop_all_mqtt_listeners(true),

    %% At this point, client reconnect and will drain
    %% their queues located at 'Node' migrating them to
    %% their new node.
    case wait_till_all_offline(Interval, N) of
        ok ->
            %% There is no guarantee that all clients will
            %% reconnect on time; we've to force migrate all
            %% offline queues.
            vmq_reg:migrate_offline_queues(TargetNodes),
            %% node is online, we'll go the proper route
            %% instead of calling leave_cluster('Node')
            %% directly
            _ = plumtree_peer_service:leave(unused_arg),
            init:stop();
        error ->
            exit("error, still online queues, check the logs, and retry!")
    end
end.
```

As it is obvious from the code, in the end if everything worked well, your VerneMQ `beam.smp` OS process should be gone. At this point you can delete the complete `data` folder as there is not much of interest anymore.

If you plan to start the same node again, we still recommend to delete the `data` folder and start from a fresh base. We're currently evaluating if keeping some of the data would help speeding up the initialization/replication process,  but for now (maybe backup) and delete the folder BEFORE starting the node again

Regarding your last question:
Is the node `vmq@vernemq-1.lan` actually offline or just not part of the cluster anymore? This shouldn't happen if the node is offline. In any case it looks like a bug.

**Comment by testark** [comment]
@dergraf many, many thanks for your thourough explanation! As for my last question, `vmq@vernemq-1.lan` is actually offline. It is taken down after running `vmq-admin cluster leave node="vmq@vernemq-1.lan" -k -i 10 -t 86400`.

**Comment by dergraf** [comment]
I've to debug this. It looks like a bug.

**Comment by larshesel** [comment]
Just took a look and this and could easily reproduce. It seems to me to be an issue here: https://github.com/erlio/plumtree/blob/master/src/plumtree_broadcast.erl#L285 where the set of peers is only updated if there are new peers, but not if any has been removed.

What do you think @dergraf ?

**Comment by ig0r** [comment]
guys, do you have any estimation for the fix?

**Comment by larshesel** [comment]
It's a trivial one as far as I can tell. The issue itself doesn't have any serious impact as plumtree is just trying to start an exchange with a node that is no longer there. It's annoying though.

**Comment by ig0r** [comment]
ok, thanks for the heads-up!

**Comment by francois-travais** [comment]
Any update on this one ?

**Comment by larshesel** [comment]
As we're moving away from plumtree in favor of the swc backend I'll close this for now.
